src/Libraries/DSCPython/CPythonEvaluator.cs

```diff
@@ -10,6 +10,119 @@
 
 namespace DSCPython
 {
+    /// <summary>
+    /// Used to comapre DynamoCPythonHandles by their PythonIDs
+    /// </summary>
+    class DynamoCPythonHandleComparer : IEqualityComparer<DynamoCPythonHandle>
+    {
+
+        public bool Equals(DynamoCPythonHandle x, DynamoCPythonHandle y)
+        {
+            if (x.PythonObjectID.Equals(y.PythonObjectID))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        public int GetHashCode(DynamoCPythonHandle obj)
+        {
+            return obj.PythonObjectID.GetHashCode();
+        }
+    }
+
+    /// <summary>
+    /// This class wraps a PythonNet.PyObj and performs
+    /// disposal tasks to make sure the underlying object is removed from
+    /// the shared global scope between all CPython scopes.
+    /// </summary>
+    [IsVisibleInDynamoLibrary(false)]
+    public class DynamoCPythonHandle : IDisposable
+    {
+        /// <summary>
+        /// A static map of DynamoCPythonHandle counts, is used to avoid removing the underlying python objects from the 
+        /// global scope while there are still handles referencing them.
+        /// </summary>
+        internal static Dictionary<DynamoCPythonHandle, int> HandleCountMap = new Dictionary<DynamoCPythonHandle, int>(new DynamoCPythonHandleComparer());
+
+        /// <summary>
+        /// A unique ID that identifies this python object. It's as a lookup
+        /// symbol within the global scope to find this python object instance.
+        /// </summary>
+        internal IntPtr PythonObjectID { get; set; }
+        /// <summary>
+        /// A string representation of this object.
+        /// TODO - it may be better to just call ToString - this was an attempt
+        /// to optimize needing to use locks and the GIL to get strings for preview bubbles.
+        /// </summary>
+        internal string StringRepresentation { get; set; }
```

Comment:
this property is used in the `ToString` override method for this class - what do you think about this given the comment above? We could take the other approach, get rid of this property, and have every call to toString make a call to python.

Reply:
I think we should probably use it. That way we don't need to deal with locking every time.

One thing I'm not seeing though, is if this is getting updated as Python nodes change the object. Could be something to test.

Reply:
yep, let me take a look at that, because we're just using the handle, and that address should not change I thought this was okay to cache for the object lifetime, but maybe it gives a false impression and could be useful for debugging to update it at some points.

Reply:
Yes, but if you update the object, then its `ToString` representation might change, I guess. It depends on what `PyObject` does.

Reply:
yes, you're right, this will get out of date if users implement `__repr__` or `__str__` on their classes. Will think on it. Might need to ditch the cacheing...

Reply:
I'm leaning towards the following.
1. for now, just remove the property, remove the cacheing logic, and let the toString lock and get the most up to date value.

2. if we find in the future that this causes performance problems worth optimizing, we can add cacheing back and update ALL existing DSCPythonHandles string values every time any Cpython engine executes, so this update occurs all within the same lock statment -  We could try to make this stricter and check for references to avoid unnecessary updates, but I'm not sure we would detect all nested properties etc.

Reply:
Option 1 sounds ok. This should be an uncommon usage of Python anyway.